File: render-layout-lambda-node-improve/renderer.js
/**
 * renderer.js - Canvas rendering functionality with improved fallback
 */

const fs = require('fs');
const { CANVAS } = require('./config');
const { log } = require('./common');
const { fetchImage, isValidImageUrl } = require('./services');
const { splitTextToLines } = require('./utils');

// Track canvas availability
let canvasAvailable = true;
let pngjs;
let canvasModule;

// Lazy-load canvas module with fallback
function getCanvasModule() {
  if (canvasModule) return canvasModule;
  
  if (!canvasAvailable) {
    log('Canvas module unavailable, using fallback renderer');
    return null;
  }
  
  try {
    canvasModule = require('canvas');
    log('Canvas module loaded successfully');
    return canvasModule;
  } catch (err) {
    log('Failed to load canvas module:', err.message);
    canvasAvailable = false;
    
    // Try to load pngjs as fallback
    try {
      pngjs = require('pngjs');
      log('Loaded pngjs as fallback renderer');
    } catch (pngErr) {
      log('Failed to load pngjs fallback:', pngErr.message);
    }
    
    return null;
  }
}

/**
 * Create fallback "canvas" when canvas module is unavailable
 */
function createFallbackCanvas(width, height) {
  log('Creating fallback canvas', { width, height });
  
  // Create a basic PNG with pngjs if available
  if (pngjs) {
    const { PNG } = pngjs;
    try {
      const png = new PNG({ width, height });
      // Fill with white
      for (let y = 0; y < height; y++) {
        for (let x = 0; x < width; x++) {
          const idx = (width * y + x) << 2;
          png.data[idx] = 255;       // R
          png.data[idx + 1] = 255;   // G
          png.data[idx + 2] = 255;   // B
          png.data[idx + 3] = 255;   // A
        }
      }
      
      return {
        width,
        height,
        getContext: () => ({
          scale: () => {},
          fillStyle: '',
          fillRect: () => {},
          fillText: () => {},
          textAlign: '',
          textBaseline: '',
          font: '',
          imageSmoothingEnabled: true,
          measureText: () => ({ width: 100 }),
          strokeStyle: '',
          lineWidth: 1,
          strokeRect: () => {},
          beginPath: () => {},
          moveTo: () => {},
          lineTo: () => {},
          stroke: () => {},
          drawImage: () => {}
        }),
        toBuffer: () => PNG.sync.write(png)
      };
    } catch (pngErr) {
      log('Error creating PNG:', pngErr.message);
    }
  }
  
  // Basic object that mimics canvas when all else fails
  return {
    width,
    height,
    getContext: () => ({
      scale: () => {},
      fillStyle: '',
      fillRect: () => {},
      fillText: () => {},
      textAlign: '',
      textBaseline: '',
      font: '',
      imageSmoothingEnabled: true,
      measureText: () => ({ width: 100 }),
      strokeStyle: '',
      lineWidth: 1,
      strokeRect: () => {},
      beginPath: () => {},
      moveTo: () => {},
      lineTo: () => {},
      stroke: () => {},
      drawImage: () => {}
    }),
    toBuffer: () => Buffer.from(''),
  };
}

/**
 * Create canvas with fallback mechanism
 */
function createSafeCanvas(width, height) {
  const canvas = getCanvasModule();
  if (!canvas) {
    return createFallbackCanvas(width, height);
  }
  
  try {
    const { createCanvas } = canvas;
    log('Creating canvas', { width, height });
    return createCanvas(width, height);
  } catch (err) {
    log('Error creating canvas:', err.message);
    canvasAvailable = false;
    return createFallbackCanvas(CANVAS.fallbackWidth, CANVAS.fallbackHeight);
  }
}

/**
 * Main render function with improved fallback
 */
async function renderLayout(layout) {
  log('renderLayout: started', { layoutId: layout.layoutId });

  try {
    // Check for canvas availability early
    getCanvasModule();
    
    // Get layout trays
    const trays = layout.subLayoutList?.[0]?.trayList || [];
    log(`Found ${trays.length} trays in layout`);

    // Check maximum tray limit
    if (trays.length > CANVAS.maxTrays) {
      throw new Error(`Layout exceeds maximum tray limit of ${CANVAS.maxTrays}`);
    }

    // Calculate canvas dimensions
    const numColumns = 7; // Fixed at 7 columns
    const numRows = Math.ceil(trays.length / numColumns) || 1;
    const { cell, row, header, footer, padding, titlePadding, metadataHeight } = CANVAS;
    
    // Limit size to prevent crashes
    const canvasWidth = Math.min(padding * 2 + numColumns * cell.width + (numColumns - 1) * cell.spacing, 2000);
    const canvasHeight = Math.min(
      padding * 2 +
      titlePadding +
      header.height +
      numRows * (cell.height + footer.height) +
      (numRows - 1) * row.spacing +
      footer.height +
      metadataHeight, 
      2000
    );

    // Create canvas with fallback support
    const canvas = createSafeCanvas(canvasWidth * CANVAS.scale, canvasHeight * CANVAS.scale);
    const ctx = canvas.getContext('2d');
    
    if (canvasAvailable) {
      ctx.scale(CANVAS.scale, CANVAS.scale);
      ctx.imageSmoothingEnabled = true;
      
      // Draw background
      ctx.fillStyle = 'white';
      ctx.fillRect(0, 0, canvasWidth, canvasHeight);
      
      // Set default text alignment
      ctx.textAlign = 'center';
      ctx.textBaseline = 'middle';
      
      // Draw components only if canvas is available
      await drawTitle(ctx, layout, canvasWidth, padding);
      await drawColumnHeaders(ctx, numColumns, padding, titlePadding, header.height);
      await drawRows(ctx, trays, numColumns, padding, titlePadding, header.height);
      await drawFooter(ctx, layout, canvasWidth, canvasHeight, padding, metadataHeight);
    }

    log('renderLayout: finished drawing, returning PNG buffer');
    return canvas.toBuffer('image/png');
  } catch (renderErr) {
    log('Critical error in renderLayout:', renderErr.message);
    return createErrorCanvas(renderErr, layout);
  }
}

/**
 * Draw the layout title
 */
async function drawTitle(ctx, layout, canvasWidth, padding) {
  if (!canvasAvailable) return;
  
  const title = `Kootoro Vending Machine Layout (ID: ${layout.layoutId || 'Unknown'})`;
  ctx.font = 'bold 18px Arial';
  ctx.fillStyle = 'black';
  ctx.fillText(title, canvasWidth / 2, padding);
}

/**
 * Draw column headers
 */
async function drawColumnHeaders(ctx, numColumns, padding, titlePadding, headerHeight) {
  if (!canvasAvailable) return;
  
  ctx.font = '14px Arial';
  ctx.fillStyle = 'black';
  
  for (let col = 0; col < numColumns; col++) {
    const x = padding + col * (CANVAS.cell.width + CANVAS.cell.spacing) + CANVAS.cell.width / 2;
    const y = padding + titlePadding + headerHeight / 2;
    ctx.fillText(`${col + 1}`, x, y);
  }
}

/**
 * Draw all rows and their slots
 */
async function drawRows(ctx, trays, numColumns, padding, titlePadding, headerHeight) {
  if (!canvasAvailable) return;
  
  for (let rowIdx = 0; rowIdx < trays.length; rowIdx++) {
    const tray = trays[rowIdx];
    const rowLetter = tray.trayCode || String.fromCharCode(65 + rowIdx);
    const rowY = padding + titlePadding + headerHeight + 
                 rowIdx * (CANVAS.cell.height + CANVAS.footer.height + CANVAS.row.spacing);

    // Draw row separator line if not first row
    if (rowIdx > 0) {
      const separatorY = rowY - CANVAS.row.spacing / 2;
      ctx.strokeStyle = 'rgb(200, 200, 200)';
      ctx.lineWidth = 1.0 / CANVAS.scale;
      ctx.beginPath();
      ctx.moveTo(padding, separatorY);
      ctx.lineTo(padding + numColumns * (CANVAS.cell.width + CANVAS.cell.spacing) - CANVAS.cell.spacing, separatorY);
      ctx.stroke();
    }

    // Draw row label
    ctx.font = '16px Arial';
    ctx.fillStyle = 'black';
    ctx.textAlign = 'right';
    ctx.fillText(rowLetter, padding - CANVAS.textPadding, rowY + CANVAS.cell.height / 2);
    ctx.textAlign = 'center';

    // Get slots for this tray and sort by slot number
    const slots = tray.slotList ? tray.slotList.sort((a, b) => a.slotNo - b.slotNo) : [];

    // Process images in smaller batches - only if canvas is available
    if (canvasAvailable) {
      // Fetch images in smaller batches to reduce memory usage
      const imagePromises = [];
      for (const slot of slots) {
        if (slot.productTemplateImage && isValidImageUrl(slot.productTemplateImage)) {
          imagePromises.push(fetchImage(slot.productTemplateImage));
        } else {
          imagePromises.push(Promise.resolve(null));
        }
      }
      
      const batchSize = 3; // Smaller batch size for ARM64
      const imageBuffers = [];
      
      for (let i = 0; i < imagePromises.length; i += batchSize) {
        const batchPromises = imagePromises.slice(i, i + batchSize);
        const batchResults = await Promise.all(batchPromises);
        imageBuffers.push(...batchResults);
      }

      // Draw each cell in the row
      for (let col = 0; col < numColumns; col++) {
        await drawCell(ctx, slots, rowLetter, col, rowY, imageBuffers);
      }
    }
  }
}

/**
 * Draw an individual cell
 */
async function drawCell(ctx, slots, rowLetter, col, rowY, imageBuffers) {
  if (!canvasAvailable) return;
  
  const slot = slots.find((s) => s.slotNo === col + 1);
  const cellX = CANVAS.padding + col * (CANVAS.cell.width + CANVAS.cell.spacing);

  // Draw cell background and border
  ctx.fillStyle = 'rgb(250, 250, 250)';
  ctx.fillRect(cellX, rowY, CANVAS.cell.width, CANVAS.cell.height);
  ctx.strokeStyle = 'rgb(180, 180, 180)';
  ctx.lineWidth = 1.0 / CANVAS.scale;
  ctx.strokeRect(cellX, rowY, CANVAS.cell.width, CANVAS.cell.height);

  if (!slot) return;
  
  const positionCode = `${rowLetter}${col + 1}`;
  
  // Draw position code
  ctx.textAlign = 'left';
  ctx.font = 'bold 14px Arial';
  ctx.fillStyle = 'rgb(0, 0, 150)';
  ctx.fillText(positionCode, cellX + 8, rowY + 16);
  ctx.textAlign = 'center';

  // Calculate image position
  const imgX = cellX + (CANVAS.cell.width - CANVAS.image.size) / 2;
  const imgY = rowY + (CANVAS.cell.height - CANVAS.image.size) / 2 - 10;
  
  // Get image buffer
  const imageBuffer = imageBuffers[slots.indexOf(slot)];
  
  if (imageBuffer) {
    try {
      const canvas = getCanvasModule();
      if (canvas) {
        const { loadImage } = canvas;
        const img = await loadImage(imageBuffer);
        ctx.drawImage(img, imgX, imgY, CANVAS.image.size, CANVAS.image.size);
      }
    } catch (err) {
      log('Failed to load image:', { error: err.message });
      drawImagePlaceholder(ctx, cellX, imgX, imgY);
    }
  } else {
    drawImagePlaceholder(ctx, cellX, imgX, imgY);
  }

  // Draw product name
  const nameY = imgY + CANVAS.image.size + 15;
  ctx.font = '12px Arial';
  ctx.fillStyle = 'black';
  
  let productName = slot.productTemplateName ? slot.productTemplateName.trim() : '';
  if (productName === '') productName = 'Sản phẩm';
  
  const maxWidth = CANVAS.cell.width - 20;
  const lines = splitTextToLines(ctx, productName, maxWidth);
  
  for (let i = 0; i < lines.length; i++) {
    const lineY = nameY + i * 18;
    ctx.fillText(lines[i], cellX + CANVAS.cell.width / 2, lineY);
  }
}

/**
 * Draw a placeholder for missing images
 */
function drawImagePlaceholder(ctx, cellX, imgX, imgY) {
  if (!canvasAvailable) return;
  
  ctx.fillStyle = 'rgb(240, 240, 240)';
  ctx.fillRect(imgX, imgY, CANVAS.image.size, CANVAS.image.size);
  ctx.strokeStyle = 'rgb(200, 200, 200)';
  ctx.lineWidth = 0.5 / CANVAS.scale;
  ctx.strokeRect(imgX, imgY, CANVAS.image.size, CANVAS.image.size);
  ctx.font = '10px Arial';
  ctx.fillStyle = 'rgb(150, 150, 150)';
  ctx.fillText('Image Unavailable', cellX + CANVAS.cell.width / 2, imgY + CANVAS.image.size / 2);
}

/**
 * Draw the footer with layout ID and timestamp
 */
async function drawFooter(ctx, layout, canvasWidth, canvasHeight, padding, metadataHeight) {
  if (!canvasAvailable) return;
  
  const footerText = `Kootoro Vending Machine Layout (ID: ${layout.layoutId || 'Unknown'})`;
  const footerY = canvasHeight - padding / 2 - metadataHeight;
  
  ctx.font = 'bold 18px Arial';
  ctx.fillStyle = 'black';
  ctx.fillText(footerText, canvasWidth / 2, footerY);

  const now = new Date();
  const formattedDate = now.toISOString().replace('T', ' ').substring(0, 19);
  const metadataText = `Generated at: ${formattedDate}`;
  
  ctx.font = '12px Arial';
  ctx.fillStyle = 'rgb(100, 100, 100)';
  ctx.fillText(metadataText, canvasWidth / 2, canvasHeight - 10);
}

/**
 * Create an error canvas when rendering fails
 */
function createErrorCanvas(error, layout) {
  try {
    // If canvas is available, create an error canvas
    const canvas = getCanvasModule();
    if (canvas && canvasAvailable) {
      const { createCanvas } = canvas;
      const errorCanvas = createCanvas(800, 400);
      const errorCtx = errorCanvas.getContext('2d');
      
      errorCtx.fillStyle = 'white';
      errorCtx.fillRect(0, 0, 800, 400);
      
      errorCtx.font = 'bold 20px Arial';
      errorCtx.fillStyle = 'red';
      errorCtx.textAlign = 'center';
      errorCtx.fillText('Error rendering layout', 400, 100);
      
      errorCtx.font = '16px Arial';
      errorCtx.fillStyle = 'black';
      errorCtx.fillText(`Layout ID: ${layout.layoutId || 'Unknown'}`, 400, 150);
      errorCtx.fillText(`Error: ${error.message}`, 400, 200);
      errorCtx.fillText(`Time: ${new Date().toISOString()}`, 400, 250);
      
      return errorCanvas.toBuffer('image/png');
    }
    
    // If canvas isn't available, use fallback method
    return createFallbackCanvas(800, 400).toBuffer();
    
  } catch (canvasErr) {
    log('Failed to create error canvas:', canvasErr.message);
    
    // Fallback to pngjs
    if (pngjs) {
      try {
        const { PNG } = pngjs;
        const png = new PNG({ width: 400, height: 200 });
        
        // Fill with white
        for (let y = 0; y < 200; y++) {
          for (let x = 0; x < 400; x++) {
            const idx = (400 * y + x) << 2;
            // White background
            png.data[idx] = 255;
            png.data[idx + 1] = 255;
            png.data[idx + 2] = 255;
            png.data[idx + 3] = 255;
            
            // Add red border
            if (x < 5 || x > 395 || y < 5 || y > 195) {
              png.data[idx] = 255;     // R
              png.data[idx + 1] = 0;   // G
              png.data[idx + 2] = 0;   // B
            }
          }
        }
        
        return PNG.sync.write(png);
      } catch (pngErr) {
        log('Failed to create error PNG:', pngErr.message);
      }
    }
    
    // Last resort - empty buffer
    return Buffer.from('');
  }
}

module.exports = {
  renderLayout
};

File: prod/api-function/api_images/upload-render-json/renderer/renderer.go
package renderer

import (
	"bytes"
	"context"
	"crypto/md5"
	"fmt"
	"image"
	"image/png"
	"io"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"api_images_upload_render/config"

	"github.com/fogleman/gg"
)

type Layout struct {
	LayoutID        int64       `json:"layoutId"`
	LayoutType      *int        `json:"layoutType,omitempty"`      // Optional field from real data
	MachineModelID  *int        `json:"machineModelId,omitempty"`  // Optional field from real data
	SubLayoutList   []SubLayout `json:"subLayoutList"`
}

type SubLayout struct {
	TrayList []Tray `json:"trayList"`
}

type Tray struct {
	LayoutTrayID *int64 `json:"layoutTrayId,omitempty"` // Optional field from real data
	TrayCode     string `json:"trayCode"`
	TrayNo       int    `json:"trayNo"`
	SlotList     []Slot `json:"slotList"`
}

type Slot struct {
	VmLayoutSlotId       int    `json:"vmLayoutSlotId"`
	ProductId            int    `json:"productId"`
	ProductTemplateId    int    `json:"productTemplateId"`
	MaxQuantity          int    `json:"maxQuantity"`
	SlotNo               int    `json:"slotNo"`
	Status               int    `json:"status"`
	Position             string `json:"position"`
	SlotIndexCode        int    `json:"slotIndexCode"`
	CellNumber           int    `json:"cellNumber"`
	ProductTemplateName  string `json:"productTemplateName"`
	ProductTemplateImage string `json:"productTemplateImage"`
}

func RenderLayoutToBytes(layout Layout) ([]byte, error) {
	cfg := config.GetConfig()

	// Check if there are any trays to render
	if len(layout.SubLayoutList) == 0 || len(layout.SubLayoutList[0].TrayList) == 0 {
		return nil, fmt.Errorf("no trays found in layout")
	}

	trays := layout.SubLayoutList[0].TrayList
	numRows := len(trays)

	// Calculate the actual number of columns needed based on the JSON data
	actualColumns := 0
	for _, tray := range trays {
		for _, slot := range tray.SlotList {
			if slot.SlotNo > actualColumns {
				actualColumns = slot.SlotNo
			}
		}
	}
	
	// Ensure we have at least 1 column and don't exceed the maximum
	if actualColumns == 0 {
		actualColumns = 1
	}
	if actualColumns > cfg.NumColumns {
		actualColumns = cfg.NumColumns
	}

	canvasWidth := cfg.Padding*2 + float64(actualColumns)*cfg.CellWidth + float64(actualColumns-1)*cfg.CellSpacing
	canvasHeight := cfg.Padding*2 + cfg.TitlePadding + cfg.HeaderHeight +
		float64(numRows)*(cfg.CellHeight+cfg.FooterHeight) +
		float64(numRows-1)*cfg.RowSpacing +
		cfg.FooterHeight + cfg.MetadataHeight

	dc := gg.NewContext(int(canvasWidth*cfg.CanvasScale), int(canvasHeight*cfg.CanvasScale))
	dc.Scale(cfg.CanvasScale, cfg.CanvasScale)

	// Set white background
	dc.SetRGB(1, 1, 1)
	dc.Clear()

	// Load fonts
	titleFont, err := gg.LoadFontFace(cfg.BoldFontPath, cfg.TitleFontSize)
	if err != nil {
		return nil, fmt.Errorf("failed to load title font: %v", err)
	}
	dc.SetFontFace(titleFont)

	// Draw title
	title := fmt.Sprintf("Kootoro Vending Machine Layout (ID: %d)", layout.LayoutID)
	dc.SetRGB(0, 0, 0)
	dc.DrawStringAnchored(title, canvasWidth/2, cfg.Padding, 0.5, 0.5)

	// Load column font
	columnFont, err := gg.LoadFontFace(cfg.FontPath, cfg.HeaderFontSize)
	if err != nil {
		return nil, fmt.Errorf("failed to load column font: %v", err)
	}
	dc.SetFontFace(columnFont)

	// Draw column numbers
	for col := 0; col < actualColumns; col++ {
		x := cfg.Padding + float64(col)*(cfg.CellWidth+cfg.CellSpacing) + cfg.CellWidth/2
		y := cfg.Padding + cfg.TitlePadding + cfg.HeaderHeight/2
		dc.DrawStringAnchored(fmt.Sprintf("%d", col+1), x, y, 0.5, 0.5)
	}

	// Draw rows
	for rowIdx, tray := range trays {
		rowLetter := tray.TrayCode
		rowY := cfg.Padding + cfg.TitlePadding + cfg.HeaderHeight + float64(rowIdx)*(cfg.CellHeight+cfg.FooterHeight+cfg.RowSpacing)

		if rowIdx > 0 {
			separatorY := rowY - cfg.RowSpacing/2
			dc.SetRGB(0.784, 0.784, 0.784)
			dc.SetLineWidth(1.0 / cfg.CanvasScale)
			dc.DrawLine(cfg.Padding, separatorY, canvasWidth-cfg.Padding, separatorY)
			dc.Stroke()
		}

		// Draw row letter
		dc.SetRGB(0, 0, 0)
		rowFont, err := gg.LoadFontFace(cfg.FontPath, 16.0)
		if err != nil {
			return nil, fmt.Errorf("failed to load row font: %v", err)
		}
		dc.SetFontFace(rowFont)
		dc.DrawStringAnchored(rowLetter, cfg.Padding-cfg.TextPadding, rowY+cfg.CellHeight/2, 1.0, 0.5)

		// Sort slots by slotNo
		sort.Slice(tray.SlotList, func(i, j int) bool {
			return tray.SlotList[i].SlotNo < tray.SlotList[j].SlotNo
		})

		// Load position font
		positionFont, err := gg.LoadFontFace(cfg.BoldFontPath, cfg.PositionFontSize)
		if err != nil {
			return nil, fmt.Errorf("failed to load position font: %v", err)
		}
		dc.SetFontFace(positionFont)

		for col := 0; col < actualColumns; col++ {
			slot := findSlotByNo(tray.SlotList, col+1)
			cellX := cfg.Padding + float64(col)*(cfg.CellWidth+cfg.CellSpacing)

			// Draw cell background
			dc.SetRGB(0.98, 0.98, 0.98)
			dc.DrawRectangle(cellX, rowY, cfg.CellWidth, cfg.CellHeight)
			dc.Fill()

			// Draw cell border
			dc.SetRGB(0.706, 0.706, 0.706)
			dc.SetLineWidth(1.0 / cfg.CanvasScale)
			dc.DrawRectangle(cellX, rowY, cfg.CellWidth, cfg.CellHeight)
			dc.Stroke()

			if slot != nil {
				// Draw position code
				positionCode := fmt.Sprintf("%s%d", rowLetter, col+1)
				dc.SetRGB(0, 0, 0.588)
				dc.DrawString(positionCode, cellX+8, rowY+16)

				// Product image
				imgX := cellX + (cfg.CellWidth-cfg.ImageSize)/2
				imgY := rowY + (cfg.CellHeight-cfg.ImageSize)/2 - 10

				// Load the image
				loadImageCtx, cancel := context.WithTimeout(context.Background(), time.Duration(cfg.ImageLoadTimeout)*time.Second)
				img, err := loadImageFromURL(loadImageCtx, slot.ProductTemplateImage)
				cancel()
				if err != nil {
					log.Printf("Failed to load image for %s: %v", slot.Position, err)
					// Draw placeholder
					dc.SetRGB(0.941, 0.941, 0.941)
					dc.DrawRectangle(imgX, imgY, cfg.ImageSize, cfg.ImageSize)
					dc.Fill()
					dc.SetRGB(0.784, 0.784, 0.784)
					dc.SetLineWidth(0.5 / cfg.CanvasScale)
					dc.DrawRectangle(imgX, imgY, cfg.ImageSize, cfg.ImageSize)
					dc.Stroke()

					// Load placeholder font
					placeholderFont, err := gg.LoadFontFace(cfg.FontPath, 10.0)
					if err == nil {
						dc.SetFontFace(placeholderFont)
					}
					dc.SetRGB(0.588, 0.588, 0.588)
					dc.DrawStringAnchored("Image Unavailable", cellX+cfg.CellWidth/2, imgY+cfg.ImageSize/2, 0.5, 0.5)
				} else {
					// Scale and draw image
					dc.Push()
					dc.Translate(imgX, imgY)
					scaleX := cfg.ImageSize / float64(img.Bounds().Dx())
					scaleY := cfg.ImageSize / float64(img.Bounds().Dy())
					dc.Scale(scaleX, scaleY)
					dc.DrawImage(img, 0, 0)
					dc.Pop()
				}

				// Draw product name
				nameY := imgY + cfg.ImageSize + 15
				productFont, err := gg.LoadFontFace(cfg.FontPath, 12.0)
				if err != nil {
					return nil, fmt.Errorf("failed to load product font: %v", err)
				}
				dc.SetFontFace(productFont)
				dc.SetRGB(0, 0, 0)
				productName := strings.TrimSpace(slot.ProductTemplateName)
				if productName == "" {
					productName = "Sản phẩm"
				}
				maxWidth := cfg.CellWidth - 20
				lines := splitTextToLines(dc, productName, maxWidth)
				for i, line := range lines {
					lineY := nameY + float64(i)*18
					dc.DrawStringAnchored(line, cellX+cfg.CellWidth/2, lineY, 0.5, 0.5)
				}
			}
		}
	}

	// Draw footer
	footerFont, err := gg.LoadFontFace(cfg.BoldFontPath, 18.0)
	if err != nil {
		return nil, fmt.Errorf("failed to load footer font: %v", err)
	}
	dc.SetFontFace(footerFont)
	footerText := fmt.Sprintf("Kootoro Vending Machine Layout (ID: %d)", layout.LayoutID)
	footerY := canvasHeight - cfg.Padding/2 - cfg.MetadataHeight
	dc.SetRGB(0, 0, 0)
	dc.DrawStringAnchored(footerText, canvasWidth/2, footerY, 0.5, 0.5)

	// Draw metadata
	metadataFont, err := gg.LoadFontFace(cfg.FontPath, 12.0)
	if err != nil {
		return nil, fmt.Errorf("failed to load metadata font: %v", err)
	}
	dc.SetFontFace(metadataFont)
	now := time.Now()
	formattedDate := now.Format("Jan 02, 2006 15:04:05")
	metadataText := fmt.Sprintf("Generated at: %s", formattedDate)
	dc.SetRGB(0.392, 0.392, 0.392)
	dc.DrawStringAnchored(metadataText, canvasWidth/2, canvasHeight-10, 0.5, 0.5)

	// Encode the image to PNG
	var buf bytes.Buffer
	img := dc.Image()
	encoder := &png.Encoder{
		CompressionLevel: png.BestSpeed,
	}
	err = encoder.Encode(&buf, img)
	if err != nil {
		return nil, fmt.Errorf("failed to encode PNG: %v", err)
	}

	return buf.Bytes(), nil
}

func findSlotByNo(slots []Slot, slotNo int) *Slot {
	for _, slot := range slots {
		if slot.SlotNo == slotNo {
			return &slot
		}
	}
	return nil
}

func splitTextToLines(dc *gg.Context, text string, maxWidth float64) []string {
	words := strings.Split(text, " ")
	if len(words) == 0 {
		return []string{""}
	}
	lines := []string{}
	currentLine := words[0]
	for _, word := range words[1:] {
		testLine := currentLine + " " + word
		w, _ := dc.MeasureString(testLine)
		if w <= maxWidth {
			currentLine = testLine
		} else {
			lines = append(lines, currentLine)
			currentLine = word
		}
	}
	if currentLine != "" {
		lines = append(lines, currentLine)
	}
	if len(lines) > 2 {
		lines[1] = lines[1][:len(lines[1])-3] + "..."
		return lines[:2]
	}
	return lines
}

func loadImageFromURL(ctx context.Context, url string) (image.Image, error) {
	if url == "" {
		return nil, fmt.Errorf("empty image URL")
	}

	cacheDir := "/tmp/image_cache"
	err := os.MkdirAll(cacheDir, 0755)
	if err != nil {
		return nil, err
	}

	hash := md5.Sum([]byte(url))
	cachedFile := filepath.Join(cacheDir, fmt.Sprintf("%x.png", hash))

	if _, err := os.Stat(cachedFile); err == nil {
		f, err := os.Open(cachedFile)
		if err != nil {
			return nil, err
		}
		defer f.Close()
		img, _, err := image.Decode(f)
		return img, err
	}

	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return nil, err
	}
	client := &http.Client{
		Timeout: 20 * time.Second,
	}

	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("HTTP status %d", resp.StatusCode)
	}
	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	err = os.WriteFile(cachedFile, data, 0644)
	if err != nil {
		log.Printf("Warning: Failed to write image to cache: %v", err)
	}
	img, _, err := image.Decode(bytes.NewReader(data))
	return img, err
}
